### PR TITLE
Gptconfig fix

### PR DESCRIPTION
**Pull request by djlisbonne**
Reworking the way GPTConfig objects are created and used for loading from pre_trained models to preserve model_args when resuming training, sampling, or finetuning from prev_runs.

### CODE REVIEW

File: train.py [hunk 1]
```diff
@@ -464,48 +465,58 @@ def setup(self):
...
             self.model = GPT(gptconf)
-            ## TODO: Add means here to udpate the resume for: block size (finetune for longer context), rotary type, rope length, softmax type, etc.
+
             ## TODO: Add ability here to swap WTE factors.
             state_dict = checkpoint['model']
             for k,v in list(state_dict.items()):
                 if k.startswith('_orig_mod.'):
                     state_dict[k[len('_orig_mod.'):]] = state_dict.pop(k)
             self.model.load_state_dict(state_dict)
-            self.iter_num = checkpoint['iter_num']
```
Comment (gkielian) [inline train.py:483]: We should keep this -- or at least the option to keep this -- in case we are continuing from an existing run with a learning rate schedule.

File: train.py [hunk 2]
```diff
@@ -464,48 +465,58 @@ def setup(self):
...
             self.best_val_loss = checkpoint['best_val_loss']
+
+            exit()
```
Comment (gkielian) [inline train.py:499]: Just checking is there a need for an exit here?

File: train.py [hunk 3]
```diff
@@ -464,48 +465,58 @@ def setup(self):
...
-            self.best_val_loss = checkpoint['best_val_loss']
+            variation_dict = model_variation_dictionary[self.args.init_from]
+            # NOTE: the hierarchy of parameters goes: 1)variation_dict >> 2)cmd-line args >> 3)GPTConfig defaults
+            for k in variation_dict:
+                gptconf[k] = variation_dict[k]
+
+            self.model = GPT.from_pretrained(gptconf, model_type=self.args.gpt2_type)
+            self.load_data()
+
+            exit()
```
Comment (gkielian) [inline train.py:519]: Curious why the exit

Comment (gkielian) [reply train.py:519]: looks good

**Review by gkielian — COMMENTED** [review]
Some feedback, we should try testing this in comparison to the legacy repo to see if there are differences.

Also thinking we might want to associate some default training parameters with different gpt's ,  we should have one default for training and one default for finetuning (can pull from the orginal repo for these numbers).

File: train.py (hunk 1 above)
Comment (djlisbonne) [reply train.py:483]: This functionality is still preserved in the new "resume"-or-"prev_run" branching, it sets iter_num equal to the checkpointed value or to 0, just as it did before.

Comment (gkielian) [reply train.py:483]: cool

**Review by gkielian — APPROVED** [review]
Looks good
